src/core/net/dns_headers.hpp

```diff
@@ -374,6 +374,243 @@ class Header : public Clearable<Header>
+     * E.g., ""_http._tcp", "_http._tcp." (same as previous one), "host-1.test".
+     *
+     * This method validates that the @p aLabels is a valid domain name format, i.e., no empty label, and labels are
+     * `kMaxLabelLength` (63) characters or less.
+     *
+     * @note This method NEVER adds a label terminator (empty label) to the message, even in the case where @p aLabels
+     * ends with a dot character, e.g., "host-1.test." is treated same as "host-1.test".
+     *
+     * @param[in]  aLabels            A domain name label string. Can be nullptr (then treated as "").
+     * @param[in]  aMessage           The message to which to append the encoded name.
+     *
+     * @retval OT_ERROR_NONE          Successfully encoded and appended the domain name label(s) to @p aMessage.
+     * @retval OT_ERROR_INVALID_ARGS  Domain name label @p aLabels is not valid.
+     * @retval OT_ERROR_NO_BUFS       Insufficient available buffers to grow the message.
+     *
+     */
+    static otError AppendMultipleLabels(const char *aLabels, Message &aMessage);
+
+    /**
+     * This static method appends a domain name label terminator to a message.
+     *
+     * An encoded domain name is terminated by an empty label (a zero byte).
+     *
+     * @param[in] aMessage            The message to append to.
+     *
+     * @retval OT_ERROR_NONE          Successfully encoded and appended the terminator label to @p aMessage.
+     * @retval OT_ERROR_NO_BUFS       Insufficient available buffers to grow the message.
+     *
+     */
+    static otError AppendTerminator(Message &aMessage);
+
+    /**
+     * This static method appends a pointer type domain name label to a message.
+     *
+     * Pointer label is used for domain name compression. It allows an entire domain name or a list of labels at the
+     * end of an encoded domain name to be replaced with a pointer to a prior occurrence of the same name within the
+     * message.
+     *
+     * @param[in] aOffset             The offset from the start of DNS header to use for pointer value.
+     * @param[in] aMessage            The message to append to.
+     *
+     * @retval OT_ERROR_NONE          Successfully encoded and appended the pointer label to @p aMessage.
+     * @retval OT_ERROR_NO_BUFS       Insufficient available buffers to grow the message.
+     *
+     */
+    static otError AppendPointerLabel(uint16_t aOffset, Message &aMessage);
+
+    /**
+     * This static method encodes and appends a full domain name to a message.
+     *
+     * The @p aName must follow  "<label1>.<label2>.<label3>", i.e., a sequence of labels separated by dot '.' char.
+     * E.g., "example.com", "example.com." (same as previous one), local.", "default.service.arpa", "." or "" (root).
+     *
+     * This method validates that the @p aName is a valid domain name format, i.e. no empty labels, and labels are
```

Comment:
Are we validating everything in RFC 1035 Section 2.3?

Reply:
No. Just the 3 things mentioned, are really checked, i.e., no empty label, labels limited to 63 chars or less, overall name 255 bytes (the first two are useful to ensure the encoded name can be parsed correctly).

We can add more checks/validations but I am not sure if it would be useful. Main downside is that it would add code-size. I think we can rely on caller/provider of the names to provide correct/valid ones.